lib/extensions/extensions.js

```diff
@@ -0,0 +1,173 @@
+ *  their own departmental frontend(Styles, Scripts, nunjucks etc)
+ *
+ * Module.exports
+ *    getPublicUrls:
+ *      Params: (type | string ) eg. 'scripts', 'stylesheets'
+ *      Description:
+ *        returns array of urls for a type (script, stylesheet, nunjucks etc).
+ *    getFileSystemPaths:
+ *      Params: (type | string ) eg. 'scripts', 'stylesheets'
+ *      Description:
+ *        returns array paths to the file in the filesystem for a type (script, stylesheet, nunjucks etc)
+ *    getPublicUrlAndFileSystemPaths:
+ *      Params: (type | string ) eg. 'scripts', 'stylesheets'
+ *      Description:
+ *        returns Array of objects, each object is an extension and each obj has the filesystem & public url for the given type
+ *    getAppConfig:
+ *      Params: (type | string ) eg. 'scripts', 'stylesheets'
+ *      Description:
+ *        Returns an object containing two keys(scripts & stylesheets), each item contains an array of full paths to specific files.
+ *        This is used in the views to output links and scripts each file.
+ *    getAppViews:
+ *      Params: (additionalViews | Array ) eg.extensions.getAppViews([path.join(__dirname, '/app/views/'),path.join(__dirname, '/lib/')])
+ *      Description:
+ *        Returns an array of paths to nunjucks templates which is used to configure nunjucks in server.js
+ *    setExtensionsByType
+ *      Params: N/A
+ *      Description: only used for test purposes to reset mocked extensions items to ensure they are up-to-date when the tests run
+ *
+ * *
+ */
+
+// Core dependencies
+const fs = require('fs')
+const path = require('path')
+
+// Local dependencies
+const appConfig = require('../../app/config')
+
+// Generic utilities
+const removeDuplicates = arr => [...new Set(arr)]
+const filterOutParentAndEmpty = part => part && part !== '..'
+const objectMap = (object, mapFn) => Object.keys(object).reduce((result, key) => {
+  result[key] = mapFn(object[key], key)
+  return result
+}, {})
+
+// File utilities
+const getPathFromProjectRoot = (...all) => {
+  return path.join.apply(null, [__dirname, '..', '..'].concat(all))
+}
+const pathToPackageConfigFile = packageName => getPathFromProjectRoot('node_modules', packageName, 'govuk-prototype-kit.config.json')
+
+const readJsonFile = (filePath) => {
+  return JSON.parse(fs.readFileSync(filePath, 'utf8'))
+}
+const getPackageConfig = packageName => {
+  if (fs.existsSync(pathToPackageConfigFile(packageName))) {
+    return readJsonFile(pathToPackageConfigFile(packageName))
+  } else {
+    return {}
+  }
+}
+
+// Handle errors to do with extension paths
+// Example of `subject`: { packageName: 'govuk-frontend', item: '/all.js' }
+const throwIfBadFilepath = subject => {
+  if (('' + subject.item).indexOf('\\') > -1) {
+    throw new Error(`Can't use backslashes in extension paths - "${subject.packageName}" used "${subject.item}".`)
+  }
+  if (!('' + subject.item).startsWith('/')) {
```

Comment:
small thing but I think `thing === false` is clearer than `!thing` - what do people think?

Reply:
It's never too late to discuss the small things IMHO.  This will be run multiple times, probably at runtime too.  If performance matters for the prototype kit in general then it matters in this instance.

Reply:
it's not currently part of our code style, so happy to come back to this later